lib/msf/core/post/process.rb

```diff
@@ -45,6 +46,28 @@ def get_processes
     end
   end
 
+  #
+  # Terminate process with ID `pid` on the remote system
+  #
+  # @return [Boolean] True upon success
+  #
+  def kill_process(pid)
+    if session.platform == 'windows'
+      if session.type == 'meterpreter'
+        session.sys.process.kill(pid)
+        return true
+      else
+        return !cmd_exec("taskkill /PID #{pid}").starts_with?('ERROR')
+      end
+    else
+      return cmd_exec("kill -9 #{pid} && echo true").to_s.include?('true')
+    end
```

Comment:
This could be updated in the future to add this command to the other Meterpreters. Instead of checking the platform and type, could you switch it to check if the `stdapi_sys_process_kill` command is available?

It should be:

```suggestion
    if session.type == 'meterpreter' && session.commands.include?(Rex::Post::Meterpreter::Extensions::Stdapi::COMMAND_ID_STDAPI_SYS_PROCESS_KILL)
      session.sys.process.kill(pid)
      return true
    elsif session.platform == 'windows'
      return !cmd_exec("taskkill /PID #{pid}").starts_with?('ERROR')
    else
      return cmd_exec("kill -9 #{pid} && echo true").to_s.include?('true')
    end
```

This would ensure that as more meterpreters add that command, it'll automatically get picked up. Since you're failing back to the OS command as well, you can remove the necessary compatibility requirement on line 12.

Reply:
Thanks. Capability detection is a superior approach. We should be doing more of this in Framework.

Elsewhere in Framework we load_extapi in case it is not already loaded. Although load error is fatal. I have a planned PR to make this less fatal, allowing fallback.

> Since you're failing back to the OS command as well, you can remove the necessary compatibility requirement on line 12.

I don't think that's true.

My understanding, perhaps incorrect, is that Meterpreter capabilities listed in a module are capabilities required for Meterpreter sessions, regardless of whether a shell command fallback with `cmd_exec` exists.

If the `stdapi_sys_process_kill` capability should be removed then presumably the existing `stdapi_sys_process_get_processes` capability should also be removed, as `session.sys.process.get_processes` is used only in the `meterpreter_get_processes` method which also includes a `cmd_exec` shell command fallback:

https://github.com/rapid7/metasploit-framework/blob/322c31f042cbd933e2ed291ed510dce165360a10/lib/msf/core/post/process.rb#L69-L75

Reply:
> My understanding, perhaps incorrect, is that Meterpreter capabilities listed in a module/library are capabilities required for Meterpreter sessions, regardless of whether a shell command fallback with cmd_exec exists.

Yeah, you're right we can probably just leave it as a requirement for now. I was going to refer to the [registry](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/registry.rb#L58) mixin as an example of one that left the reqs out because each method had a command equivalent but it too marks everything as required. If it becomes a problem, we'll probably just want to update them all at once.

Til then, using the API though should be more quite as it won't execute an external command.